README.md

```diff
@@ -102,114 +24,93 @@ import { wait } from 'f-promise-async';
 The funnel can be closed with `fun.close()`.  
-When a funnel is closed, the operations that are still in the funnel will continue but their callbacks
+When a funnel is closed, the operations that are still in the funnel will continue but their callbacks 
 won't be called, and no other operation will enter the funnel.
 
 ### handshake and queue
 
-* `hs = fpromise.handshake()`  
+* `hs = handshake()`  
   allocates a simple semaphore that can be used to do simple handshakes between two tasks.  
   The returned handshake object has two methods:  
-  `hs.wait()`: waits until `hs` is notified.  
-  `hs.notify()`: notifies `hs` (without waiting for an acknowledgement)
+  `await hs.wait()`: waits until `hs` is notified.  
+  `hs.notify()`: notifies `hs`.  
   Note: `wait` calls are not queued. An exception is thrown if wait is called while another `wait` is pending.
-* `q = fpromise.queue(options)`  
+
+* `q = new Queue(options)`  
   allocates a queue which may be used to send data asynchronously between two tasks.  
   The `max` option can be set to control the maximum queue length.  
-  When `max` has been reached `q.put(data)` discards data and returns false.
+  When `max` has been reached `q.put(data)` discards data and returns false.  
   The returned queue has the following methods:  
-  `data = q.read()`: dequeues an item from the queue. Waits if no element is available.  
-  `q.write(data)`:  queues an item. Waits if the queue is full.  
+  `data = await q.read()`:  dequeue and returns the first item. Waits if the queue is empty. Does not allow concurrent read.  
+  `await q.write(data)`:  queues an item. Waits if the queue is full.  
   `ok = q.put(data)`: queues an item synchronously. Returns true if the queue accepted it, false otherwise.  
-  `q.end()`: ends the queue. This is the synchronous equivalent of `q.write(undefined)`  
+  `q.end()`: ends the queue. This is the synchronous equivalent of `q.write(_, undefined)`  
   `data = q.peek()`: returns the first item, without dequeuing it. Returns `undefined` if the queue is empty.  
   `array = q.contents()`: returns a copy of the queue's contents.  
-  `q.adjust(fn[, thisObj])`: adjusts the contents of the queue by calling `newContents = fn(oldContents)`.  
-  `q.length`: number of items currently in the queue.  
- 
-### CLS (Continuation Local Storage)
+  `q.adjust(fn[, thisObj])`: adjusts the contents of the queue by calling `newContents = fn(oldContents)`.
 
-* `cx = fpromise.context()`  
-  returns the current context.
+### Continuation local storage (CLS)
 
-* `fn = fpromise.withContext(fn, cx)`  
-  wraps a function so that it executes with context `cx` (or a wrapper around current context if `cx` is falsy).
+* `result = await withContext(fn, cx)`  
+  wraps a function so that it executes with context `cx` (or a wrapper around current context if `cx` is falsy).  
   The previous context will be restored when the function returns (or throws).  
   returns the wrapped function.
```

Comment:
It is the only tool that need to be executed in a `run`. Can be a nested function call (coroutine).

Reply:
done

Reply:
\` autour du `run` stp

Reply:
done